kep/108-Memory-bandwidth-and-LLC-aware-filtering-and-scoring/README.md

```diff
@@ -0,0 +1,160 @@
+
+## Goals
+1. Use scheduler plugin, which is the most Kubernetes native way, to implement memory bandwidth and LLC aware filtering and scoring.
+2. Leverage memory bandwidth and LLC relative metrics for filtering and scoring.
+3. Provide configurable weights for prioritizing the metrics used in the scoring calculations.
+
+## Non-Goals
+
+## Use Cases
+1. If a workload is memory bandwidth bond, e.g. rpmbuild, it uses much memory bandwidth. If too much such kind of workloads are scheduled to a node, the memory bandwith of the node will be used out quickly. The performance of the workloads will drop significantly. If the scheduler uses this filter plugin it can be avoid. 
+
+2. If there are many workloads run on a node, the LLC contention may become very intense. The performance of all workloads will drop. If the scheduler uses this score plugin, it can ensure new workload is scheduled to LLC freer node. Furtherly, mitigate performance drop.
+
+## Terms
+- LLC: last level cache.
+- RDT: Intel® Resource Director Technology.
+- SLA: Service Level Agreement.
+- MPKI: Misses Per Kilo Instructions.
+
+## Proposal
+We introduce some metrics such as memory bandwidth utilization, LLC occupation etc. At schedule stage, we can leverage these metrics to assist filtering and scoring. It is hard to determine how important a resource is and how much resource is required. We did off-line analysis to different scenario and get a series of coefficient of correlation and generic usage value. The coefficient is called affinity and the generic usage value is called profile. The affinities and profiles are provided by a yaml configuration file. Detail about how the affinity and profile are used is described in the following design details section. The average value of a period of time is more meaningful than immediate value for some metrics, for instance, LLC occupancy and CPU usage. Metrics data aren't obtained from agents (such as cAdvisor) directly, instead they are stored in Prometheus first and then are retrieved on demand.
+
+## Design Details
+The filtering and scoring plugin leverages the metrics retrieved from Prometheus to determine the node candidates for running a pod. The metrics considered by the plugin at the moment are:
+
+* For filtering out oversubscribe nodes: 
+
+    * free memory available (GB)
```

Comment:
.spec.request.memory may not be specified or the specified value may not be proper, it is for such kind of cases.

Reply:
.spec.request.memory is the KSM (kubernetes resource model) we have to follow, otherwise, the pod may behave inconsistly among different k8s components (scheduler, kubelet) and external integrators.

Reply:
So generally the idea of proposing a Score plugin makes sense to me, but not the idea of introducing a Filter plugin.